pkg/repositories/transformers/task_execution.go

```diff
@@ -240,6 +242,94 @@ func mergeCustom(existing, latest *_struct.Struct) (*_struct.Struct, error) {
 	return &response, nil
 }
 
+// mergeExternalResource combines the lastest ExternalResourceInfo proto with an existing instance
+// by updating fields and merging logs.
+func mergeExternalResource(existing, latest *event.ExternalResourceInfo) *event.ExternalResourceInfo {
+	if existing == nil {
+		return latest
+	}
+
+	if latest == nil {
+		return existing
+	}
+
+	if latest.ExternalId != "" && existing.ExternalId != latest.ExternalId {
+		existing.ExternalId = latest.ExternalId
+	}
+	// we only update if the index is equal so updating existing.Index is not necessary
+	if latest.RetryAttempt != 0 && existing.RetryAttempt != latest.RetryAttempt {
+		existing.RetryAttempt = latest.RetryAttempt
+	}
+	existing.Phase = latest.Phase
+	if latest.CacheStatus != core.CatalogCacheStatus_CACHE_DISABLED && existing.CacheStatus != latest.CacheStatus {
+		existing.CacheStatus = latest.CacheStatus
+	}
+	existing.Logs = mergeLogs(existing.Logs, latest.Logs)
+
+	return existing
+}
+
+// mergeExternalResources combines lists of external resources. This involves appending new
+// resources and updating in-place resources attributes.
+func mergeExternalResources(existing, latest []*event.ExternalResourceInfo) []*event.ExternalResourceInfo {
+	if len(latest) == 0 {
+		return existing
+	}
+
+	for i, externalResource := range latest {
+		var index int
+		if externalResource.GetIndex() == 0 && externalResource.GetRetryAttempt() == 0 && externalResource.GetPhase() == 0 {
+			// updating an external resource to retryAttempt 0 and phase 0 is invalid, because that
+			// is the initial state, so we use this condition to signify an external resource
+			// without these parameters (ex. a simple ExternalResourceID).
+			index = i
+		} else {
+			index = int(externalResource.GetIndex())
+		}
+
+		if index >= len(existing) {
+			// if the latest external resources contains an out of order update
+			// (ie. index > len(existing)) then we should append placeholder external resources
+			// that will be updated later
+			for index-1 >= len(existing) {
+				existing = append(existing, &event.ExternalResourceInfo{Index: uint32(len(existing))})
+			}
+
+			existing = append(existing, externalResource)
+		} else {
+			existing[index] = mergeExternalResource(existing[index], externalResource)
+		}
+	}
+
+	return existing
+}
+
+// mergeMetadata merges an existing TaskExecutionMetadata instance with the provided instance. This
+// includes updating non-defaulted fields and merging ExternalResources.
+func mergeMetadata(existing, latest *event.TaskExecutionMetadata) *event.TaskExecutionMetadata {
+	if existing == nil {
+		return latest
+	}
+
+	if latest == nil {
+		return existing
+	}
+
+	if latest.GeneratedName != "" && existing.GeneratedName != latest.GeneratedName {
```

Comment:
if I'm understanding correctly, latest can never explicitly unset any fields (i.e. set them to their zero value)?

Reply:
I tried to implement logical merging with the ability to unset fields that make sense to unset. For example, `GeneratedName` is set to the task ID, `PluginIdentifier` is the plugin name, and `InstanceClass` is whether it is interruptible or not. To me, it doesn't make sense to allow null values for these, so in hopes to (potentially) reduce event construction complexity we allow updates to have nulls. I think it makes sense for some fields to be updatable to default values, and tried to allow it. For example the `Phase` in the mergeExternalResource function. I am not tied to this logic. If you, or anybody, feels strongly about it, I'm happy to update. Thoughts?

Reply:
this sounds reasonable, i think in other events we take the same approach (merging log links, etc) - thanks for explaining!